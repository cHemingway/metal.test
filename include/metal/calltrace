/**
 * @file   metal/test/calltrace
 * @date   02.05.2017
 * @author Klemens D. Morgenstern
 *



 This header automatically selects either the C or C++ version of the header, depending on that compiler setttings.
 */


#if defined(__cplusplus)
#include <metal/calltrace.hpp>
#else
#include <metal/calltrace.h>
#endif