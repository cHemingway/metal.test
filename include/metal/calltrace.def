/**
 * @file   metal/test/calltrace.def
 * @date   01.05.2017
 * @author Klemens D. Morgenstern
 *
 */

struct metal_calltrace_
{
    const void * fn;
    
    const void ** content;
    const int content_size;
    
    const int repeat;
    const int skip;
        
	int to_skip;
    int errored;
    int repeated;
    int current_position;
    int start_depth;
};

#if defined(ULLONG_MAX)
typedef unsigned long long metal_timestamp_t;
#else
typedef unsigned long metal_timestamp_t;
#endif

metal_timestamp_t metal_timestamp() __attribute__((no_instrument_function));

int __metal_set_calltrace  (struct metal_calltrace_ * ct) __attribute__((no_instrument_function));
int __metal_reset_calltrace(struct metal_calltrace_ * ct) __attribute__((no_instrument_function));