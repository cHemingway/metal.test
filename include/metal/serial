#if defined(__cplusplus)
#include <metal/serial.hpp>
#else
#include <metal/serial.h>
#endif