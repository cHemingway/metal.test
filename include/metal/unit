#if defined(__cplusplus)
#include <metal/unit.hpp>
#else
#include <metal/unit.h>
#endif