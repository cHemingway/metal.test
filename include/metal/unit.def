/**
 * @file   metal/unit.def
 * @date   25.10.2016
 * @author Klemens D. Morgenstern
 *
 */

#ifndef METAL_UNIT_DEF_
#define METAL_UNIT_DEF_

typedef enum __metal_level_t
{
    __metal_level_assert,
    __metal_level_expect,
} __metal_level;

typedef enum __metal_oper_t
{
    __metal_oper_enter_case,
    __metal_oper_exit_case,
    __metal_oper_enter_ranged,
    __metal_oper_exit_ranged,
    __metal_oper_log,
    __metal_oper_checkpoint,
    __metal_oper_message,
    __metal_oper_plain,
    __metal_oper_predicate,
    __metal_oper_equal,
    __metal_oper_not_equal,
    __metal_oper_close,
    __metal_oper_close_rel,
    __metal_oper_close_perc,
    __metal_oper_ge,
    __metal_oper_greater,
    __metal_oper_le,
    __metal_oper_lesser,
    __metal_oper_exception,
    __metal_oper_any_exception,
    __metal_oper_no_exception,
    __metal_oper_no_exec,
    __metal_oper_exec,
    __metal_oper_report
} __metal_oper;

extern int __metal_status  ;
extern int __metal_critical;
extern int __metal_errored;

void __metal_call(void (*func)(), const char * msg,
               const char * file, int line);

void __metal_unit_break(__metal_level lvl,
               __metal_oper oper,
               int condition,
               int bitwise,
               const char* str1,
               const char* str2,
               const char* str3,
               const char* file,
               int line);

#if defined(METAL_UNIT_ONLY_ERRORS)
#define __metal_unit_impl(lvl, oper, condition, ...) (((condition == 0) && (lvl == __metal_level_assert)) ? __metal_unit_break(lvl, oper, condition, __VA_ARGS__) : ((void)0))
#elif defined(METAL_UNIT_ONLY_WARNINGS)
#define __metal_unit_impl(lvl, oper, condition, ...) ((condition == 0) ? __metal_unit_break(lvl, oper, condition, __VA_ARGS__) : ((void)0))
#else
#define __metal_unit_impl(...) __metal_unit_break(__VA_ARGS__)
#endif

#if defined(METAL_UNIT_ONLY_ERRORS) || defined(METAL_UNIT_ONLY_WARNINGS)
#define __metal_unit_call(Function, ...) Function()
#else
#define __metal_unit_call(...) __metal_call(__VA_ARGS__)
#endif

int __metal_report();

#define __METAL_BITWISE_STEP(Lhs, Rhs, Index, Oper) (((Lhs >> Index) & 1) Oper ((Rhs >> Index) & 1))

#define __METAL_BITWISE_8(Lhs, Rhs, Oper, Chain) \
    ((Lhs & 1) Oper (Rhs & 1))                   \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x4, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x5, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x6, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x7, Oper)

#define __METAL_BITWISE_16(Lhs, Rhs, Oper, Chain) \
    __METAL_BITWISE_8(Lhs, Rhs, Oper, Chain)        \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x8, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x9, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xA, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xB, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xC, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xD, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xE, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xF, Oper)

#define __METAL_BITWISE_32(Lhs, Rhs, Oper, Chain) \
    __METAL_BITWISE_16(Lhs, Rhs, Oper, Chain)        \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x10, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x11, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x12, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x13, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x14, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x15, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x16, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x17, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x18, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x19, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1A, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1B, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1C, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1D, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1E, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1F, Oper)

#define __METAL_BITWISE_64(Lhs, Rhs, Oper, Chain)    \
    __METAL_BITWISE_32(Lhs, Rhs, Oper, Chain)        \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x20, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x21, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x22, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x23, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x24, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x25, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x26, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x27, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x28, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x29, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2A, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2B, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2C, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2D, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2E, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2F, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x30, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x31, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x32, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x33, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x34, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x35, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x36, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x37, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x38, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x39, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3A, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3B, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3C, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3D, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3E, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3F, Oper)

#define __METAL_BITWISE_SIZE(Lhs, Rhs) ((sizeof(Lhs) >= sizeof(Rhs)) ? sizeof(Lhs) : sizeof(Rhs))

#define METAL_CONCAT_IMPL( x, y ) x##y
#define METAL_CONCAT( x, y ) METAL_CONCAT_IMPL( x, y )

#define METAL_STRINGIZE(x) METAL_STRINGIZE2(x)
#define METAL_STRINGIZE2(x) #x
#define METAL_LOCATION_STR() __FILE__ "(" METAL_STRINGIZE(__LINE__) ")"

#define __METAL_PP_NARG(...)  __METAL_PP_NARG_(__VA_ARGS__,__METAL_PP_RSEQ_N())
#define __METAL_PP_NARG_(...) __METAL_PP_ARG_N(__VA_ARGS__)
#define __METAL_PP_ARG_N(_1, _2, _3, _4, _5, _6, _7, _8, _9,_A, _B, _C, _D, _E, _F, N,...) N
#define __METAL_PP_RSEQ_N() F, E, D, C, B, A, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0

#define METAL_OVERLOAD(Macro, Args...) METAL_CONCAT(Macro, __METAL_PP_NARG(Args)) (Args)

#endif
