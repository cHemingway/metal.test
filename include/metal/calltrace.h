/**
 * @file   metal/test/calltrace.h
 * @date   02.05.2017
 * @author Klemens D. Morgenstern
 *




 This header provides the C version of the test macros.
 */
#ifndef METAL_CALLTRACE_H_
#define METAL_CALLTRACE_H_

#include <limits.h>
#include <metal/calltrace.def>

#if !defined(METAL_CALLTRACE_DOXYGEN)

#define METAL_NO_INSTRUMENT __attribute__((no_instrument_function))

#endif

#if defined(METAL_CALLTRACE_DOXYGEN)

///The function to be implemented if a timestamp shall be added to the function calls.
metal_timestamp_t metal_timestamp();

///The unsigned int type that is returned by the timestamp function.
typedef detail::metal_timestamp_t metal_timestamp_t;


/**The type carrying the necessary information for the calltrace.
 */
struct metal_calltrace
{
    ///The function which's subcalls are trace.
    const void * fn;

    ///The content, i.e. array of the calls
    const void ** content;
    ///The amount of functions traced
    const int content_size;

    ///How many function calls shall be traced
    const int repeat;
    ///How many function calls shall be ignored before tracing
    const int skip;
};

#else


typedef struct metal_calltrace_ metal_calltrace;

#endif

/**
 This function initializes and registers a calltrace.
 Note that there is a limited amount of calltraces that can be added. The default value is 16, but can be changed by defining
 `METAL_CALLTRACE_STACK_SIZE` while compiling calltrace.

 \section metal_calltrace_init_example Example
 \code{.cpp}

 #include <cassert>
 #include <metal/calltrace.h>

 void func() {foo(); bar();}

 int main(int argc, char *argv[])
 {
    const void * ct_arr[] = {&foo, &bar};
    metal_calltrace ct = {&func, ct_arr, 2, 0, 0};

    assert(metal_calltrace_init(&ct));

    func();
    assert(metal_calltrace_success(&ct));
    assert(metal_calltrace_deinit(&ct));
    return 0;
 }
 \endcode

 \param ct A pointer to the calltrace to be initialized.
 \return A value differnt from zero if it succeeded.

 */
static int METAL_NO_INSTRUMENT metal_calltrace_init(metal_calltrace * ct)
{
    ct->to_skip = ct->skip;
    ct->errored = 0;
    ct->repeated = 0;
    ct->current_position = 0;
    ct->start_depth = -1;

    return __metal_set_calltrace(ct);
}


/** This functions returns a value different from zero if the calltrace was completed.
 * The behaviour depends on the repeat setting of the calltrace.
 * If the calltrace is set to repeat n-times it has to be at least repeated once,
 * while any other number will require the calltrace to be repeated exactly as set.
 *
 * @param ct A pointer to the calltrace that shall be examined.
 * @return Unequal zero if successful.
 */
static int METAL_NO_INSTRUMENT metal_calltrace_complete(metal_calltrace * ct)
{
    int rep_res = ct->repeat == 0 ? (ct->repeated > 0 ) : (ct->repeated >= ct->repeat);
    return (ct->current_position == 0) && rep_res;
}


/** This functions returns a value different from zero if the calltrace was completed and did not err.
 *
 * @param ct A pointer to the calltrace that shall be examined.
 * @return Unequal zero if completed without error.
 */
static int METAL_NO_INSTRUMENT metal_calltrace_success (metal_calltrace * ct)
{
    return metal_calltrace_complete(ct) && !ct->errored;
}

/** This function deinitializes the calltrace and removes it from the stack.
 * @warning This function must be called before the calltrace struct is removed from the stack, otherwise it might casue memory leaks.
 *
 * @param ct The calltrace to be deinitialized
 * @return A value unequal to zero if the calltrace was succesful removed from the stack.
 */
static int METAL_NO_INSTRUMENT metal_calltrace_deinit(metal_calltrace * ct)
{
    return __metal_reset_calltrace(ct);
}


#endif /* METAL_TEST_CALLTRACE_H_ */
