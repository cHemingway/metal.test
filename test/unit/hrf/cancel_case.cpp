
#include <metal/unit>

void func()
{
    METAL_CRITICAL(METAL_ASSERT(false));
}

int main(int argc, char * argv[])
{
    METAL_CALL(func, "my test case");
    METAL_EXPECT(false);
    return METAL_REPORT();
}
