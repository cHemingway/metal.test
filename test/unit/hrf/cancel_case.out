starting test execution
cancel_case.cpp(11) entering test case [my test case]
cancel_case.cpp(6) critical assertion failed [expression]: false
    canceling test case [my test case]: { executed : 1, warnings : 0, errors : 1}
cancel_case.cpp(12) expectation failed [expression]: false
free tests : { executed : 1, warnings : 1, errors : 0}
full test report: { executed : 2, warnings : 1, errors : 1}